File: SpringBoot-SpringData/SpringData/springbootStudy/src/main/java/com/itdragon/service/UserService.java
package com.itdragon.service;

import javax.transaction.Transactional;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import com.itdragon.common.ItdragonResult;
import com.itdragon.pojo.User;
import com.itdragon.repository.UserRepository;

@Service
@Transactional
public class UserService {
	
	@Autowired
	private UserRepository userRepository;
	
    public ItdragonResult registerUser(User user) {
    	// 检查用户名是否注册，一般在前端验证的时候处理，因为注册不存在高并发的情况，这里再加一层查询是不影响性能的
    	if (null != userRepository.findByAccount(user.getAccount())) {
    		return ItdragonResult.build(400, "");
    	}
    	userRepository.save(user);
    	// 注册成功后选择发送邮件激活。现在一般都是短信验证码
    	return ItdragonResult.build(200, "");
    }
    
    public ItdragonResult editUserEmail(String email) {
    	// 通过Session 获取用户信息, 这里假装从Session中获取了用户的id，后面讲解SOA面向服务架构中的单点登录系统时，修改此处代码 FIXME
    	long id = 3L;
    	// 添加一些验证，比如短信验证
    	userRepository.updateUserEmail(id, email);
    	return ItdragonResult.ok();
    }
    
}

File: SpringBoot/spring-boot-springsecurity-jwt/src/main/kotlin/com/itdragon/server/user/service/UserService.kt
package com.itdragon.server.user.service

import com.itdragon.server.user.domain.ITDragonUser
import com.itdragon.server.user.model.UserCreateInfo
import com.itdragon.server.user.model.UserUpdateInfo
import com.itdragon.server.user.repository.UserRepository
import org.springframework.beans.BeanUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.domain.Specification
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import javax.annotation.Resource
import javax.persistence.EntityExistsException
import javax.persistence.EntityNotFoundException

@Service
class UserService {
    @Autowired
    lateinit var userRepository: UserRepository

    @Autowired
    lateinit var passwordEncoder: PasswordEncoder

    fun createUser(userCreateInfo: UserCreateInfo): ITDragonUser {
        userRepository.findByUsername(userCreateInfo.username)?.let {
            throw EntityExistsException("用户名[${userCreateInfo.username}] 已经存在")
        }
        userCreateInfo.password = passwordEncoder.encode(userCreateInfo.password)
        val user = ITDragonUser()
        BeanUtils.copyProperties(userCreateInfo, user)
        return userRepository.save(user)
    }

    fun updateUser(userUpdateInfo: UserUpdateInfo): ITDragonUser {
        val existed = userRepository.findById(userUpdateInfo.id).get()
        BeanUtils.copyProperties(userUpdateInfo, existed)
        return userRepository.save(existed)
    }

    fun deleteUser(username: String): Boolean {
        val existed = findByUsername(username)
        userRepository.delete(existed)
        return true
    }

    fun listUsers(search: String?, pageable: Pageable): Page<ITDragonUser> {
        var querySpecification = Specification<ITDragonUser> { _, _, cb -> cb.conjunction() }
        search?.let {
            querySpecification = querySpecification.and { root, _, criteriaBuilder ->
                criteriaBuilder.like(root.get<String>("username"), "%$search%")
            }
        }
        return userRepository.findAll(querySpecification, pageable)
    }

    fun findByUsername(username: String): ITDragonUser {
        return userRepository.findByUsername(username)
                ?: throw EntityNotFoundException("通过用户名[$username]没有找到用户")
    }

}